Sort nested elements

Some things that are sorted can have other elements contained within them. For example:

``` cabal
build-depends: p:{l1, l2}
mixins: p(M1 as A1, M2 as A2)
-- probably other fields as well
```

Currently those interior elements are not sorted. I think that they should be.

It seems that you can't really clobber names. If you try to do something like this:

``` cabal
mixins: base ( Control.Exception, Control.Exception.Base as Control.Exception )
```

That is, intentionally try to create an alias to a module that already exists. The order doesn't matter here. It also doesn't matter if you use an existing module name or another alias. In other words, these behave similarly:

``` cabal
mixins: base ( Control.Exception.Base as Control.Exception, Control.Exception )
```

``` cabal
mixins: base ( Control.Exception as E, Control.Exception.Base as E )
```

The end result when you actually try to _use_ one of these modules is an ambiguity error from GHC:

```
.../cabal-gild/dist-newstyle/build/aarch64-osx/ghc-9.8.1/cabal-gild-1.1.4.0/build/autogen/Paths_cabal_gild.hs:15:1: error: [GHC-45102]
    Ambiguous module name ‘Control.Exception’.
    it is bound as base-4.19.0.0:Control.Exception.Base by a package flag
    it is bound as base-4.19.0.0:Control.Exception by a package flag
   |
15 | import qualified Control.Exception as Exception
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

So I think it's safe to rearrange the list of module names.

--- a/cabal-gild.cabal
+++ b/cabal-gild.cabal
@@ -121,6 +121,7 @@ library
     CabalGild.Type.Leniency
     CabalGild.Type.Line
     CabalGild.Type.List
+    CabalGild.Type.Mixin
     CabalGild.Type.Mode
     CabalGild.Type.ModuleReexport
     CabalGild.Type.Optional

--- a/source/library/CabalGild/Action/FormatFields.hs
+++ b/source/library/CabalGild/Action/FormatFields.hs
@@ -11,6 +11,7 @@ import qualified CabalGild.Type.Extension as Extension
 import qualified CabalGild.Type.ForeignLibOption as ForeignLibOption
 import qualified CabalGild.Type.Language as Language
 import qualified CabalGild.Type.LegacyExeDependency as LegacyExeDependency
+import qualified CabalGild.Type.Mixin as Mixin
 import qualified CabalGild.Type.ModuleReexport as ModuleReexport
 import qualified CabalGild.Type.PkgconfigDependency as PkgconfigDependency
 import qualified CabalGild.Type.SomeParsecParser as SPP
@@ -22,7 +23,6 @@ import qualified Distribution.FieldGrammar.Newtypes as Newtypes
 import qualified Distribution.Fields as Fields
 import qualified Distribution.ModuleName as ModuleName
 import qualified Distribution.Parsec as Parsec
-import qualified Distribution.Types.Mixin as Mixin
 import qualified Text.PrettyPrint as PrettyPrint
 
 -- | A wrapper around 'field' to allow this to be composed with other actions.

--- a/source/library/CabalGild/Type/Mixin.hs
+++ b/source/library/CabalGild/Type/Mixin.hs
@@ -1,0 +1,46 @@
+module CabalGild.Type.Mixin where
+
+import qualified Data.List as List
+import qualified Data.Ord as Ord
+import qualified Distribution.Parsec as Parsec
+import qualified Distribution.Pretty as Pretty
+import qualified Distribution.Types.IncludeRenaming as IncludeRenaming
+import qualified Distribution.Types.Mixin as Mixin
+import qualified Distribution.Types.ModuleRenaming as ModuleRenaming
+
+-- | This is a wrapper around 'Mixin.Mixin' that sorts the nested module names
+-- when parsing.
+newtype Mixin = Mixin
+  { unwrap :: Mixin.Mixin
+  }
+  deriving (Eq, Show)
+
+instance Ord Mixin where
+  compare = Ord.comparing unwrap
+
+instance Parsec.Parsec Mixin where
+  parsec = Mixin . sortMixin <$> Parsec.parsec
+
+instance Pretty.Pretty Mixin where
+  pretty = Pretty.pretty . unwrap
+
+sortMixin :: Mixin.Mixin -> Mixin.Mixin
+sortMixin x =
+  Mixin.Mixin
+    { Mixin.mixinPackageName = Mixin.mixinPackageName x,
+      Mixin.mixinLibraryName = Mixin.mixinLibraryName x,
+      Mixin.mixinIncludeRenaming = sortIncludeRenaming $ Mixin.mixinIncludeRenaming x
+    }
+
+sortIncludeRenaming :: IncludeRenaming.IncludeRenaming -> IncludeRenaming.IncludeRenaming
+sortIncludeRenaming x =
+  IncludeRenaming.IncludeRenaming
+    { IncludeRenaming.includeProvidesRn = sortModuleRenaming $ IncludeRenaming.includeProvidesRn x,
+      IncludeRenaming.includeRequiresRn = sortModuleRenaming $ IncludeRenaming.includeRequiresRn x
+    }
+
+sortModuleRenaming :: ModuleRenaming.ModuleRenaming -> ModuleRenaming.ModuleRenaming
+sortModuleRenaming x = case x of
+  ModuleRenaming.DefaultRenaming -> ModuleRenaming.DefaultRenaming
+  ModuleRenaming.HidingRenaming ys -> ModuleRenaming.HidingRenaming $ List.sort ys
+  ModuleRenaming.ModuleRenaming ys -> ModuleRenaming.ModuleRenaming $ List.sort ys

--- a/source/test-suite/Main.hs
+++ b/source/test-suite/Main.hs
@@ -669,6 +669,11 @@ main = Hspec.hspec . Hspec.parallel . Hspec.describe "cabal-gild" $ do
       "cabal-version: 2.2\nlibrary\n build-depends: p"
       "cabal-version: 2.2\n\nlibrary\n  build-depends: p\n"
 
+  Hspec.it "sorts sub-libraries in build-depends" $ do
+    expectGilded
+      "cabal-version: 3.0\nlibrary\n build-depends: p:{b,a}"
+      "cabal-version: 3.0\n\nlibrary\n  build-depends: p:{a, b}\n"
+
   Hspec.it "removes duplicate options" $ do
     -- This is kind of silly because there's only one possible foreign library
     -- option.
@@ -796,6 +801,26 @@ main = Hspec.hspec . Hspec.parallel . Hspec.describe "cabal-gild" $ do
       "library\n mixins: q (M, N as O), p"
       "library\n  mixins:\n    p,\n    q (M, N as O)\n"
 
+  Hspec.it "sorts hiding in mixins" $ do
+    expectGilded
+      "library\n mixins: p hiding (N, M)"
+      "library\n  mixins: p hiding (M, N)\n"
+
+  Hspec.it "sorts modules in mixins" $ do
+    expectGilded
+      "library\n mixins: p (N, M)"
+      "library\n  mixins: p (M, N)\n"
+
+  Hspec.it "sorts hiding in mixins requires" $ do
+    expectGilded
+      "library\n mixins: p requires hiding (N, M)"
+      "library\n  mixins: p requires hiding (M, N)\n"
+
+  Hspec.it "sorts modules in mixins requires" $ do
+    expectGilded
+      "library\n mixins: p requires (N, M)"
+      "library\n  mixins: p requires (M, N)\n"
+
   Hspec.it "does not sort ghc-options" $ do
     expectGilded
       "library\n ghc-options: b a"
@@ -830,6 +855,11 @@ main = Hspec.hspec . Hspec.parallel . Hspec.describe "cabal-gild" $ do
     expectGilded
       "custom-setup\n setup-depends: b, a"
       "custom-setup\n  setup-depends:\n    a,\n    b\n"
+
+  Hspec.it "sorts sub-libraries in setup-depends" $ do
+    expectGilded
+      "cabal-version: 3.0\ncustom-setup\n setup-depends: p:{b,a}"
+      "cabal-version: 3.0\n\ncustom-setup\n  setup-depends: p:{a, b}\n"
 
   Hspec.it "discovers an exposed module" $ do
     expectDiscover

